homeassistant/components/izone/discovery.py

```diff
@@ -0,0 +1,133 @@
+
+from .climate import ControllerDevice
+from .constants import DATA_ADD_ENTRIES, DATA_CONFIG, DATA_DISCOVERY_SERVICE
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_start_discovery_service(hass: HomeAssistantType):
+    """Set up the pizone internal discovery."""
+    class DiscoveryService(pizone.Listener):
+        """Discovery data and interfacing with pizone library."""
+
+        def __init__(self):
+            self.controllers = {}  # type: Dict[str, pizone.Controller]
+            self.controller_ready = Event()
+            self.components = {}  # type:  Dict[pizone.Controller, 'IZoneController']
+
+            self.pi_disco = None
+            self.stop_listener = None
+
+        # Listener interface
+        def controller_disconnected(self, ctrl: pizone.Controller,
+                                    ex: Exception) -> None:
+            """Disconnected from contrller."""
+            component = self.components.get(ctrl)
+            if not component:
+                return
+            component.set_available(False)
+
+        def controller_reconnected(self, ctrl: pizone.Controller) -> None:
+            """Reconnected to controller."""
+            component = self.components.get(ctrl)
+            if not component:
+                return
+            component.set_available(True)
+
+        async def _controller_discovered(self, ctrl: pizone.Controller):
+            _LOGGER.debug("Controller discovered uid=%s", ctrl.device_uid)
+
+            conf: ConfigType = hass.data.get(DATA_CONFIG)
+
+            # Filter out any entities excluded in the config file
+            if conf and ctrl.device_uid in conf[CONF_EXCLUDE]:
+                return
+
+            self.controllers[ctrl.device_uid] = ctrl
+            self.controller_ready.set()
+
+            # This will be present if the component is configured.
+            # otherwise init_controller will be called when the config entry
+            # is created.
+            async_add_entries = hass.data.get(DATA_ADD_ENTRIES)
+            if async_add_entries:
+                self.init_controller(ctrl, async_add_entries)
+
+        def init_controller(self, controller: pizone.Controller,
+                            async_add_entries):
+            """Register the controller device and the containing zones."""
+            device = ControllerDevice(controller, async_add_entries)
+            self.components[controller] = device
+
+        def controller_discovered(self, ctrl: pizone.Controller) -> None:
+            assert ctrl.device_uid not in self.controllers, \
+                "discovered device that already exists"
+
+            hass.async_create_task(self._controller_discovered(ctrl))
+
+        def controller_update(self, ctrl: pizone.Controller) -> None:
+            component = self.components.get(ctrl)
+            if not component:
+                return
+
+            hass.async_add_job(component.async_update_ha_state)
```

Comment:
Don't store the entity objects directly - use the dispatcher to send internal events around.

Reply:
I've pushed the listener stuff back into the end entities now, so no dispatching needed.